File: algorithms/102/go/main.go
package main

import "fmt"

/*
	Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

	For example:
	Given binary tree [3,9,20,null,null,15,7],
		3
	   / \
	  9  20
		/  \
	   15   7
	return its level order traversal as:
	[
	  [3],
	  [9,20],
	  [15,7]
	]
*/

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pingPong(ping, pong *[]*TreeNode) (level []int) {
	for _, node := range *ping {
		level = append(level, node.Val)
		if node.Left != nil {
			*pong = append(*pong, node.Left)
		}
		if node.Right != nil {
			*pong = append(*pong, node.Right)
		}
	}
	// clear element
	*ping = (*ping)[:0]
	return level
}

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return nil
	}

	var res [][]int
	var ping, pong []*TreeNode
	// put root in ping buf, kick it start
	ping = append(ping, root)
	isPing := true
	for len(ping) > 0 || len(pong) > 0 {
		var level []int
		if isPing {
			isPing = false
			level = pingPong(&ping, &pong)
		} else {
			isPing = true
			level = pingPong(&pong, &ping)
		}
		res = append(res, level)
	}
	return res
}

func main() {
	//[3,9,20,null,null,15,7]
	tree := &TreeNode{3, &TreeNode{9, nil, nil},
		&TreeNode{20, &TreeNode{15, nil, nil},
			&TreeNode{7, nil, nil}}}
	fmt.Println(levelOrder(tree))
}

File: algorithms/102/c++/main.cc
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;
/**
 * Definition for a binary tree node.
 */

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  void traverse(TreeNode *root, vector<vector<int>> &vals, int level) {
    if (root == nullptr) return;
    if (vals.size() == level) vals.push_back({});

    vals[level].push_back(root->val);
    traverse(root->left, vals, level + 1);
    traverse(root->right, vals, level + 1);
  }

  vector<vector<int>> levelOrder(TreeNode *root) {
    vector<vector<int>> vals;
    traverse(root, vals, 0);
    return vals;
  }
};

int main(int argc, char *argv[]) {
  Solution so;

  TreeNode n3(3);
  TreeNode n9(9);
  TreeNode n20(20);
  TreeNode n15(15);
  TreeNode n7(7);

  n3.left = &n9;
  n3.right = &n20;
  n20.left = &n15;
  n20.right = &n7;

  so.levelOrder(&n3);
  return 0;
}